Clarify Promise.prototype.finally()

allan-bonadio opened the pull request:
#### Summary
Made examples 2 & 3 more clear.
Found a number of other small things that could be improved or clarified.

#### Motivation
- First sentence described a detail rather than help with the main description.
- clarified syntax code
- clarified that finally() returns an identical promise
- removed some text that was just confusing

#### Supporting details

#### Related issues

#### Metadata
- [ ] Adds a new document
- [ x] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error

<!-- 👷‍♀️ After submitting, review the results of the "Checks" tab! -->

allan-bonadio committed:
examples 2 & 3 were unclear, etc

examples 2 & 3 were unclear, whether the resolved/rejected values were passed to the returned promise, or to the arguments to the callbacks.

Also clarified other parts which were unclear, including the first sentence, which wasn't about the main reason for finally(), but simply a detail.

allan-bonadio committed:
a few typos and add 99

allan-bonadio committed:
finally callback return

Removed references to the value returned in the finally callback.  My experiments show that if you return a rejected promise, it WILL propagate.

teoli2003 commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -13,7 +13,7 @@ browser-compat: javascript.builtins.Promise.finally
 {{JSRef}}
 
-The **`finally()`** method returns a {{jsxref("Promise")}}.
+The **`finally()`** method is called on a {{jsxref("Promise")}}.
```
```suggestion
The **`finally()`** method of a {{jsxref("Promise")}} schedules a function, the _callback function_, to be called when the promise is settled. Like `then()` and `catch()`, it immediately returns an equivalent {{jsxref("Promise")}} object, allowing to chain calls to a promise methods, an operation called _composition.
```

allan-bonadio replied:
changed to almost the same

teoli2003 commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -28,8 +28,8 @@ This helps to avoid duplicating code in both the promise's {{jsxref("Promise.the
-p.finally(function() {
-   // settled (fulfilled or rejected)
+p.finally(function onFinally() {
+   // will run after p is settled (fulfilled or rejected)
```
```suggestion
   // Code that will run after p is settled (fulfilled or rejected)
```

allan-bonadio replied:
I adopted that wording, better

teoli2003 commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -40,8 +40,9 @@ p.finally(function() {
 
-Returns a {{jsxref("Promise")}} whose `finally` handler is set to the
-specified function, `onFinally`.
+Passes through an equivalent {{jsxref("Promise")}} to the one this call was made on.
+If the function callback throws an error, the return promise will be rejected 
+with that value instead.
```
This change is incorrect I think. What about
```suggestion
Returns an equivalent {{jsxref("Promise")}} with its `finally` handler set to the specified function.
```

allan-bonadio replied:
done

teoli2003 commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -53,22 +54,23 @@ The `finally()` method is very similar to calling
-    `Promise.reject(3).finally(() => {})` will be rejected with
-    `3`.
+  fulfillment value, and so there's no need to provide it. 
+- A `finally` call will chain through an equivalent to the orignal promise.  
```
```suggestion
- A `finally` call will chain through an equivalent to the original promise.  
```

allan-bonadio replied:
doh!  done

teoli2003 reviewed (changes requested):
This is already very good. I added some suggestions/corrections.

allan-bonadio committed:
correct behavior for finally()

finally got the correct behavior for finally() if it returns a rejected promise or if it throws.

OnkarRuikar reviewed (commented):
Forgot to to click complete. Ignore if not relevant anymore.

allan-bonadio commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -28,8 +28,8 @@ This helps to avoid duplicating code in both the promise's {{jsxref("Promise.the
 
-p.finally(function() {
-   // settled (fulfilled or rejected)
+p.finally(function onFinally() {
```
I went with the arrow function

allan-bonadio commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
````diff
@@ -28,8 +28,8 @@ This helps to avoid duplicating code in both the promise's {{jsxref("Promise.the
 ```js
 p.finally(onFinally);
````
done

allan-bonadio commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -28,8 +28,8 @@ This helps to avoid duplicating code in both the promise's {{jsxref("Promise.the
 
-p.finally(function() {
-   // settled (fulfilled or rejected)
+p.finally(function onFinally() {
```
I changed it to an arrow function as per suggestion above

allan-bonadio commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -28,8 +28,8 @@ This helps to avoid duplicating code in both the promise's {{jsxref("Promise.the
-p.finally(function() {
-   // settled (fulfilled or rejected)
+p.finally(function onFinally() {
+   // will run after p is settled (fulfilled or rejected)
```
... will run after promise is settled...

allan-bonadio commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -53,22 +54,23 @@ The `finally()` method is very similar to calling
+- A `finally` call will chain through an equivalent to the orignal promise.  
+  So for example:
+  - Unlike `Promise.resolve(2).then(() => 77, () => {})` (which
+    will return a promise which will be resolved with `77`), 
```
yeah, that's better wording.  used it.

allan-bonadio commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -53,22 +54,23 @@ The `finally()` method is very similar to calling
+  - Unlike `Promise.resolve(2).then(() => 77, () => {})` (which
+    will return a promise which will be resolved with `77`), 
+    `Promise.resolve(2).finally(() => {})` will return a 
+    promise which will be resolved with `2`.
```
done

allan-bonadio commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -53,22 +54,23 @@ The `finally()` method is very similar to calling
+    `Promise.resolve(2).finally(() => {})` will return a 
+    promise which will be resolved with `2`.
+  - Similarly, unlike `Promise.reject(3).then(() => {}, () => 88)`
+    (which will, again, return a promise which will be rejected with `88`),
```
done.  It's good I'm learning the lingo.

allan-bonadio commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -53,22 +54,23 @@ The `finally()` method is very similar to calling
+    promise which will be resolved with `2`.
+  - Similarly, unlike `Promise.reject(3).then(() => {}, () => 88)`
+    (which will, again, return a promise which will be rejected with `88`),
+    `Promise.reject(3).finally(() => {})` will return a promise 
```
done

allan-bonadio commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -53,22 +54,23 @@ The `finally()` method is very similar to calling
 
 > **Note:** A `throw` (or returning a rejected promise) in the
-> `finally` callback will reject the new promise with the rejection reason
+> `finally` callback will reject the promise, which will, in that case, 
```
will reject the returned promise with the reason `99`.

allan-bonadio commented on files/en-us/web/javascript/reference/global_objects/promise/finally/index.md:
```diff
@@ -53,22 +54,23 @@ The `finally()` method is very similar to calling
+  - Similarly, unlike `Promise.reject(3).then(() => {}, () => 88)`
+    (which will, again, return a promise which will be rejected with `88`),
+    `Promise.reject(3).finally(() => {})` will return a promise 
+    which will be rejected with `3`.
```
yup

allan-bonadio committed:
Merge branch 'main' into patch-3

teoli2003 commented:
Hi! Thanks for updating the PR. I don't exactly what's going on, but I don't see your changes. Have you pushed them here?

allan-bonadio committed:
Merge branch 'main' into patch-3

allan-bonadio committed:
Incorporated suggestions

with minor differences

allan-bonadio committed:
Merge branch 'main' into patch-3

allan-bonadio committed:
Merge branch 'main' into patch-3

allan-bonadio commented:
> Hi! Thanks for updating the PR. I don't exactly what's going on, but I don't see your changes. Have you pushed them here?

sorry, never did a PR on github before.  Too many buttons and distractions, kindof confusing.  I updated the wrong branch.  Now it should be right.

teoli2003 reviewed (approved):
No worry, Git and GitHub feel different at the beginning, but work well once we are used to them.

This looks good to me. If we find more changes to be done, we can do this in follow-up PRs.

This is much better.

teoli2003 commented:
\o/
